### PR TITLE
Created initial implementation of creating model from powerflows YAML. Resolved #46

### LINKED ISSUE #46: Create initial implementation of creating model from powerflows YAML.

### PR DESCRIPTION
**Pull request by stlhrt**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -31,9 +31,24 @@
...
+            <artifactId>snakeyaml</artifactId>
+            <version>${snakeyaml.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
+            <version>${lombok.version}</version>
+            <scope>provided</scope>
+        </dependency>
         <!-- Tests -->
```
Comment (mariuszkumor) [inline pom.xml]: What about one empty line separator to keep better readability of the dependencies groups?

File: src/main/java/org/powerflows/dmn/engine/model/decision/field/Output.java [hunk 2]
```diff
@@ -30,7 +28,6 @@
     private String name;
     private String description;
     private ValueType type;
-    private Expression expression;
```
Comment (mariuszkumor) [inline src/main/java/org/powerflows/dmn/engine/model/decision/field/Output.java]: Why did you remove expression from output?

File: src/main/java/org/powerflows/dmn/io/DecisionReader.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+package org.powerflows.dmn.io;
+
+import org.powerflows.dmn.engine.model.decision.Decision;
+
+import java.io.InputStream;
+import java.util.List;
+
+public interface DecisionReader {
+
+    List<Decision> readAll(InputStream inputStream);
```
Comment (mariuszkumor) [inline src/main/java/org/powerflows/dmn/io/DecisionReader.java]: what about the param name decisionInputStream ?

File: src/main/java/org/powerflows/dmn/io/DecisionReader.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+ */
+
+package org.powerflows.dmn.io;
+
+import org.powerflows.dmn.engine.model.decision.Decision;
+
+import java.io.InputStream;
+import java.util.List;
+
+public interface DecisionReader {
```
Comment (mariuszkumor) [inline src/main/java/org/powerflows/dmn/io/DecisionReader.java]: What about DecisionReadable?

File: src/main/java/org/powerflows/dmn/io/DecisionToExternalModelConverter.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.powerflows.dmn.io;
+
+import org.powerflows.dmn.engine.model.decision.Decision;
+
+public interface DecisionToExternalModelConverter<T> {
```
Comment (mariuszkumor) [inline src/main/java/org/powerflows/dmn/io/DecisionToExternalModelConverter.java]: What about DecisionToExternalModelConvertable?

File: src/main/java/org/powerflows/dmn/io/yaml/CustomPropertyUtils.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+import org.yaml.snakeyaml.introspector.MissingProperty;
+import org.yaml.snakeyaml.introspector.Property;
+import org.yaml.snakeyaml.introspector.PropertyUtils;
+
+import java.util.Map;
+
+public class CustomPropertyUtils extends PropertyUtils {
+    @Override
+    public Property getProperty(Class<? extends Object> type, String name, BeanAccess bAccess) {
+        Map<String, Property> properties = getPropertiesMap(type, bAccess);
```
Comment (mariuszkumor) [inline src/main/java/org/powerflows/dmn/io/yaml/CustomPropertyUtils.java]: Missing a few finals in the method

Comment (mariuszkumor) [inline src/main/java/org/powerflows/dmn/io/yaml/CustomPropertyUtils.java]: I'd love to have full names of params like beanAccess instead of bAccess for better readability

File: src/main/java/org/powerflows/dmn/io/yaml/CustomPropertyUtils.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+            }
+        }
+        if (property == null && isSkipMissingProperties()) {
+            property = new MissingProperty(name);
+        }
+        if (property == null) {
+            throw new YAMLException(
+                    "Unable to find property '" + name + "' on class: " + type.getName());
+        }
+        return property;
```
Comment (mariuszkumor) [inline src/main/java/org/powerflows/dmn/io/yaml/CustomPropertyUtils.java]: One empty line before?

File: src/main/java/org/powerflows/dmn/io/yaml/CustomPropertyUtils.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+            if (property != null) {
+                properties.put(name, property);
+            }
+        }
+        if (property == null && isSkipMissingProperties()) {
+            property = new MissingProperty(name);
+        }
+        if (property == null) {
+            throw new YAMLException(
+                    "Unable to find property '" + name + "' on class: " + type.getName());
```
Comment (mariuszkumor) [inline src/main/java/org/powerflows/dmn/io/yaml/CustomPropertyUtils.java]: Is this wrapping needed? The line looks pretty short.

File: src/main/java/org/powerflows/dmn/io/yaml/CustomPropertyUtils.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+    private String toCamelCase(final String name) {
+        final String[] parts = name.split("-");
+        if (parts.length > 1) {
+            final StringBuilder sb = new StringBuilder(parts[0]);
+            for (int i = 1; i < parts.length; i++) {
+                sb.append(capitalize(parts[i]));
+            }
+
+            return sb.toString();
```
Comment (mariuszkumor) [inline src/main/java/org/powerflows/dmn/io/yaml/CustomPropertyUtils.java]: Sonar doesn't line multi-return code.

File: src/main/java/org/powerflows/dmn/io/yaml/YamlDecsionConverter.java [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+package org.powerflows.dmn.io.yaml;
+
+import org.powerflows.dmn.engine.model.decision.Decision;
+import org.powerflows.dmn.io.DecisionToExternalModelConverter;
+import org.powerflows.dmn.io.yaml.model.YamlDecision;
+
+public class YamlDecsionConverter implements DecisionToExternalModelConverter<YamlDecision> {
+    @Override
+    public YamlDecision to(final Decision decision) {
+        return null;
```
Comment (mariuszkumor) [inline src/main/java/org/powerflows/dmn/io/yaml/YamlDecsionConverter.java]: I suggest for next time throws UnsupportedOperationException, even if this is temporary.

File: src/main/java/org/powerflows/dmn/io/yaml/model/field/YamlFields.java [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+package org.powerflows.dmn.io.yaml.model.field;
+
+import lombok.Data;
+
+import java.util.Map;
+
+@Data
+public final class YamlFields {
+    private Map<String, YamlInput> in;
```
Comment (mariuszkumor) [inline src/main/java/org/powerflows/dmn/io/yaml/model/field/YamlFields.java]: Maybe inputs and outputs to get plural?

File: src/main/java/org/powerflows/dmn/io/yaml/model/rule/YamlRule.java [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+package org.powerflows.dmn.io.yaml.model.rule;
+
+import lombok.Data;
+
+import java.util.Map;
+
+@Data
+public final class YamlRule {
+    private String description;
+    private Map<String, YamlRuleInput> in;
```
Comment (mariuszkumor) [inline src/main/java/org/powerflows/dmn/io/yaml/model/rule/YamlRule.java]: Here like above?

File: src/test/groovy/org/powerflows/dmn/engine/model/evaluation/result/DecisionResultSpec.groovy [hunk 13]
```diff
@@ -23,14 +23,14 @@ class DecisionResultSpec extends Specification {
 
     void 'should build decision result and be only collection rules result'() {
         given:
-        final String someRule1Entry1Name = 'Some Rule 1 Entry 1 Name'
+        final String someRule1Entry1Name = 'Some YamlRule 1 Entry 1 Name'
```
Comment (mariuszkumor) [inline src/test/groovy/org/powerflows/dmn/engine/model/evaluation/result/DecisionResultSpec.groovy]: Why YamlRule? This test is a generic. Now suggests is dedicated for Yaml.

File: src/test/groovy/org/powerflows/dmn/io/yaml/YamlReaderSpec.groovy [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+
+    void 'should read decision using InputStream'() {
+        given:
+        final InputStream inputStream = this.class.getResourceAsStream(singleDecision)
+
+        when:
+        final Decision result = new YamlReader().read(inputStream)
+
+        then:
+        result != null
```
Comment (mariuszkumor) [inline src/test/groovy/org/powerflows/dmn/io/yaml/YamlReaderSpec.groovy]: Maybe add  result.size() == 1 ?

File: src/test/resources/org/powerflows/dmn/io/yaml/test_multiple_decisions.yml [hunk 15]
```diff
@@ -0,0 +1,109 @@
+id: sample_decision_1
```
Comment (mariuszkumor) [inline src/test/resources/org/powerflows/dmn/io/yaml/test_multiple_decisions.yml]: Is it required to keep in org/powerflows/dmn/io/yaml/ ? Or is it temporary solution?

File: README.md [hunk 16]
```diff
@@ -15,31 +15,32 @@ Power Flows supports the model in the following formats:
 ## YAML file
 
 ```yaml
-id: sample_decision_id
-name: Sample Decision Name
+id: sample_decision_1
```
[the remark this answers is not in the document]
Comment (stlhrt) [reply README.md]: reverted

File: README.md [hunk 17]
```diff
@@ -15,31 +15,32 @@ Power Flows supports the model in the following formats:
...
+name: Sample Decision One
```
[the remark this answers is not in the document]
Comment (stlhrt) [reply README.md]: reverted

File: README.md [hunk 18]
```diff
@@ -15,31 +15,32 @@ Power Flows supports the model in the following formats:
...
+      type: BOOLEAN
       description: We expect a decision, if we have access to do it
 rules:
   - description: 3 allows always
     in:
       age: 3
     out:
       allow: true
-  - in:
+  - description: 8 and red always allow
```
[the remark this answers is not in the document]
Comment (stlhrt) [reply README.md]: it's not required :)
removed

File: README.md [hunk 19]
```diff
@@ -52,19 +53,19 @@ rules:
...
     in:
       colour:
-        expression-type: feel
-        expresion: '''not(
-          "red",
-          "pink"
-        )'''
+        expression-type: FEEL
+        expression: not(
+            "red",
```
[the remark this answers is not in the document]
Comment (stlhrt) [reply README.md]: what spaces? ordinary yaml multi-line string
intendation up to field is removed automaticaly

File: pom.xml [hunk 20]
```diff
@@ -31,9 +31,24 @@
         <spock.version>1.1-groovy-2.4</spock.version>
         <cglib.version>3.2.5</cglib.version>
         <groovy.version>2.4.15</groovy.version>
+        <snakeyaml.version>1.21</snakeyaml.version>
+        <lombok.version>1.18.2</lombok.version>
+        <commons.version></commons.version>
```
[the remark this answers is not in the document]
Comment (stlhrt) [reply pom.xml]: true, cleaned

File: src/main/java/org/powerflows/dmn/io/yaml/model/field/YamlFields.java (hunk 11 above)
Comment (stlhrt) [reply src/main/java/org/powerflows/dmn/io/yaml/model/field/YamlFields.java]: it will be to verbose imho... especialy needless in not exposed API

File: src/test/groovy/org/powerflows/dmn/engine/model/evaluation/result/DecisionResultSpec.groovy (hunk 13 above)
Comment (stlhrt) [reply src/test/groovy/org/powerflows/dmn/engine/model/evaluation/result/DecisionResultSpec.groovy]: Idea did this :dagger:

File: src/main/java/org/powerflows/dmn/engine/model/decision/field/Output.java (hunk 2 above)
Comment (stlhrt) [reply src/main/java/org/powerflows/dmn/engine/model/decision/field/Output.java]: output declaration doesnt support it, its evaluated, it only makes sense to have expression in output entry inside rule

File: src/main/java/org/powerflows/dmn/io/DecisionReader.java (hunk 4 above)
Comment (stlhrt) [reply src/main/java/org/powerflows/dmn/io/DecisionReader.java]: negative! ;)

File: src/main/java/org/powerflows/dmn/io/DecisionReader.java (hunk 3 above)
Comment (stlhrt) [reply src/main/java/org/powerflows/dmn/io/DecisionReader.java]: shorter is better imho, its obvious this IS contains decision

File: src/main/java/org/powerflows/dmn/io/DecisionToExternalModelConverter.java (hunk 5 above)
Comment (stlhrt) [reply src/main/java/org/powerflows/dmn/io/DecisionToExternalModelConverter.java]: nah... as per our conversation about  ```*able```

File: src/main/java/org/powerflows/dmn/io/yaml/CustomPropertyUtils.java (hunk 6 above)
Comment (stlhrt) [reply src/main/java/org/powerflows/dmn/io/yaml/CustomPropertyUtils.java]: right

Comment (stlhrt) [reply src/main/java/org/powerflows/dmn/io/yaml/CustomPropertyUtils.java]: as above, I don't want to mess with snakeyaml code to much

File: src/main/java/org/powerflows/dmn/io/yaml/CustomPropertyUtils.java (hunk 7 above)
Comment (stlhrt) [reply src/main/java/org/powerflows/dmn/io/yaml/CustomPropertyUtils.java]: as above

File: src/main/java/org/powerflows/dmn/io/yaml/CustomPropertyUtils.java (hunk 8 above)
Comment (stlhrt) [reply src/main/java/org/powerflows/dmn/io/yaml/CustomPropertyUtils.java]: as above

File: src/test/groovy/org/powerflows/dmn/io/yaml/YamlReaderSpec.groovy (hunk 14 above)
Comment (mariuszkumor) [reply src/test/groovy/org/powerflows/dmn/io/yaml/YamlReaderSpec.groovy]: Hehe, right :)

File: src/main/java/org/powerflows/dmn/io/yaml/CustomPropertyUtils.java (hunk 9 above)
Comment (stlhrt) [reply src/main/java/org/powerflows/dmn/io/yaml/CustomPropertyUtils.java]: same as above

File: src/main/java/org/powerflows/dmn/io/yaml/YamlDecsionConverter.java (hunk 10 above)
Comment (stlhrt) [reply src/main/java/org/powerflows/dmn/io/yaml/YamlDecsionConverter.java]: ok

File: src/main/java/org/powerflows/dmn/io/yaml/CustomPropertyUtils.java (hunk 6 above)
Comment (mariuszkumor) [reply src/main/java/org/powerflows/dmn/io/yaml/CustomPropertyUtils.java]: However for now is in our code and we want to keep on clean code. That's why I preffer to make a clean up.
